index.js

```diff
@@ -47,51 +47,247 @@ const player = {
+  return Math.max(...idArray) + 1
+}
+
+function getIdsArrayFromObjArray(objectArr) {
+  return objectArr.map((obj) => obj.id)
+}
+
+function checkIfObjectIdTaken(id, getFunction) {
+  // getFunction is a function expression. we will know where to search depending on the function
+  try {
+    getFunction(id) // remember: getFunction can be: [getSong or getPlaylist] and throws Error if no song was found.
+    throw new Error('id already exist')
+  } catch (error) {
+    /* 
+    Only if the error is because of "finding" the song
+    then we continue and throw an Error.
+    */
+    if (error.message === 'id already exist') {
+      throw new Error('id already exist')
+    }
+  }
+}
+
+const durationReducer = (totalDuration, songId) => totalDuration + getSong(songId).duration
+
+function searchByQueryInSongs(query) {
+  return searchByQueryInArr(query, player.songs).sort(compareFunction('title'))
+}
+
+function searchByQueryInPlaylist(query) {
+  return searchByQueryInArr(query, player.playlists).sort(compareFunction('name'))
+}
+
+function searchByQueryInArr(query, arr) {
+  let matchArr = []
+  for (let obj of arr) {
+    for (let property of Object.values(obj)) {
+      if (('' + property).toLowerCase().includes(query)) {
+        matchArr.push(obj)
+        break
+      }
+    }
+  }
+  return matchArr
+}
+
+function compareFunction(byProperty) {
+  // returns a compareFunction based on what property to sort by, e.g "title" or "name".
+  // remember: we are sorting objects.
+  return (a, b) => {
+    if (a[byProperty] < b[byProperty]) {
+      return -1
+    }
+    if (a[byProperty] > b[byProperty]) {
+      return 1
+    }
+    return 0
+  }
+}
+
+function getClosestSong(duration) {
+  /* In this function we compare the distance between the given
+  duration and the songs duration. every time, we keep the closest song */
+  let closestSong = player.songs[0]
+  for (let song of player.songs) {
+    if (Math.abs(closestSong.duration - duration) >= Math.abs(song.duration - duration)) {
+      closestSong = song
+    }
+  }
+  return closestSong
+}
+
+function getClosestPlaylist(duration) {
+  /* Same as in getClosestSong but this time, to get the playlist
+  duration, we use the function playlistDuration */
+  let closestPlaylist = player.playlists[0]
+  for (let playlist of player.playlists) {
+    if (
+      Math.abs(playlistDuration(closestPlaylist.id) - duration) >=
+      Math.abs(playlistDuration(playlist.id) - duration)
+    ) {
+      closestPlaylist = playlist
+    }
+  }
+  return closestPlaylist
+}
+
+// end of help functions. ---------
+
 function playSong(id) {
-  // your code here
+  player.playSong(getSong(id))
 }
 
 function removeSong(id) {
-  // your code here
+  removeSongFromSongs(id)
+  removeSongFromPlaylists(id)
 }
 
-function addSong(title, album, artist, duration, id) {
-  // your code here
+function addSong(title,album,artist,duration,id = generateNewIdInArrayOfObjects(player.songs)) {
```

Comment:
Great use of default parameter values!
However, I would recommend that you put this inside the function. I, for one, am unsure if this is evaluated only once when the function is declared, or every time the function is called (and I would bet on the first option). It might mean that your "generated" value is actually always the same...